runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -6937,16 +6937,57 @@ J9::Z::TreeEvaluator::VMifInstanceOfEvaluator(TR::Node * node, TR::CodeGenerator
    return NULL;
    }
 
+TR::Register *
+J9::Z::TreeEvaluator::generateCheckForValueTypeMonitorEnterOrExit(TR::Node *node, TR::Register *objReg, TR::LabelSymbol* mergeLabel, TR::LabelSymbol *helperCallLabel, TR::CodeGenerator *cg)
+   {
+   TR::Register *objectClassReg = cg->allocateRegister();
+   TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, objectClassReg, generateS390MemoryReference(objReg, TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);
+   TR::Register *classFlagsRegister = cg->allocateRegister();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->comp()->fe());
+   generateRXInstruction(cg, TR::InstOpCode::L, node, classFlagsRegister, generateS390MemoryReference(objectClassReg, static_cast<uint32_t>(fej9->getOffsetOfClassFlags()), cg));
+   bool generateOOLSection = helperCallLabel == NULL;
+   if (generateOOLSection)
+      helperCallLabel = generateLabelSymbol(cg);
+   generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CL, node, classFlagsRegister, static_cast<int32_t>(J9ClassIsValueType), TR::InstOpCode::COND_BE, helperCallLabel);
+
+   // TODO: As we have quite amount of work happening in OOL in other cases (Monitor Cache Lookup / Lock Reservation / Regular Code Recursive CAS code for monitor),
```

Comment:
@fjeremic  I have a big-fat TODO here, as whenever we are making substantial changes in monent/monexit or work towards enabling lock reservation, we do need to refactor both the evaluator. 
Right now We have multiple OOL sections generated. 

1. MonitorCacheLookup : We implement this in OOL and as last resource there we call the JIT Helper. 
2. Lock Reservation : Also we perform some checks in OOL for lock reservation before calling the JIT helper.
3. InlineRecursive:  Generating a manual nested monitor handling in OOL, which also has a helper call.

Right now `2` is disabled and we generate either `1`(In case we do not have lockword offset) or `3` and if we generate `1` , second OOL only contains a call to JIT Helper, otherwise, we have some good work done in OOL section for `3` before it call helper.

Reply:
What you have here is fine. Personally I'd avoid complicating things with `if (generateOOLSection)` as it relies on external code doing something, so if any changes are made there would be bugs. Personally I'd unconditionally generate the OOL and call the helper just for our own sanity. However what you have is ok. I agree the evaluator is starting to become quite complicated now and needs some love.

Reply:
What I intended to do here, is if, user has provided the `helperCallLabel` I am going to use that to generate a call to helper. If we have not generated a helper call (When we are either doing inline Recursive lock routine or monitor cache lookup) , this API will generate OOL.